src/main/java/org/jabref/model/openoffice/rangesort/RangeSortVisual.java

```diff
@@ -0,0 +1,154 @@
+package org.jabref.model.openoffice.rangesort;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.jabref.model.openoffice.uno.NoDocumentException;
+import org.jabref.model.openoffice.uno.UnoScreenRefresh;
+
+import com.sun.star.awt.Point;
+import com.sun.star.lang.WrappedTargetException;
+import com.sun.star.text.XTextDocument;
+import com.sun.star.text.XTextRange;
+import com.sun.star.text.XTextViewCursor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Sort XTextRange values visually (top-down,left-to-right).
+ *
+ * Requires functional XTextViewCursor.
+ *
+ * Problem: for multicolumn layout and when viewing pages side-by-side in LO, the
+ *          (top-down,left-to-right) order interpreted as-on-the-screen: an XTextRange at the top of
+ *          the second column or second page is sorted before an XTextRange at the bottom of the
+ *          first column of the first page.
+ */
+public class RangeSortVisual {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RangeSortVisual.class);
+
+    /**
+     * Sort the input {@code inputs} visually.
+     *
+     * Requires a functional {@code XTextViewCursor}.
+     *
+     * @return The input, sorted by the elements XTextRange and getIndexInPosition.
+     */
+    public static <T> List<RangeSortable<T>> visualSort(List<RangeSortable<T>> inputs,
+                                                        XTextDocument doc,
+                                                        FunctionalTextViewCursor fcursor)
+        throws
+        WrappedTargetException,
+        NoDocumentException {
+
+        final int inputSize = inputs.size();
+
+        if (UnoScreenRefresh.hasControllersLocked(doc)) {
+            final String msg = "visualSort: with ControllersLocked, viewCursor.gotoRange is probably useless";
+            LOGGER.warn(msg);
+            throw new IllegalStateException(msg);
+        }
+
+        XTextViewCursor viewCursor = fcursor.getViewCursor();
+
+        // find coordinates
+        List<Point> positions = new ArrayList<>(inputSize);
+        for (RangeSortable<T> v : inputs) {
+            positions.add(findPositionOfTextRange(v.getRange(), viewCursor));
+        }
+        fcursor.restore(doc);
+
+        if (positions.size() != inputSize) {
+            throw new IllegalStateException("visualSort: positions.size() != inputSize");
+        }
```

Comment:
This cannot happen? One element is added to `positions` for each element in `inputs`? Or is it thrown if there are changes in the document before the method finishes?

Reply:
I would say it should not happen.
I think at some point I was losing positions corresponding to some inputs because
`set` below was a `Set` and reference marks in footnotes mapped to
the exact same coordinates (of the footnote marker), and the Set silently
ignored the duplicates.

UnoScreenRefresh.hasControllersLocked(doc) was another instance: in this case
the cursor position was not updated, and I think we got the same (wrong) position for
each range we wanted to test.

> This cannot happen? 

I hope it cannot. If it does, we have a problem.

> Or is it thrown if there are changes in the document before the method finishes?

I cannot tell what could happen if the user changes the document or even just the cursor position
while we are trying to use it. Neither the documentation of [gotoRange](https://www.openoffice.org/api/docs/common/ref/com/sun/star/text/XTextCursor.html#gotoRange) nor of [getPosition](https://www.openoffice.org/api/docs/common/ref/com/sun/star/text/XTextViewCursor.html#getPosition) discusses the possibility of failure. Nor does the [getViewCursor](https://www.openoffice.org/api/docs/common/ref/com/sun/star/text/XTextViewCursorSupplier.html#getViewCursor) documentation mentions it. Yet they can, and did fail,
and I was trying to program against *observed* (as opposed to *documented*) behaviour.

On the other hand we have two more checks on the size, probably the last one is enough to discover if there is a problem.

Removing the first two tests, keeping only `if (result.size() != inputSize)`